File: bbsync/api/v1/parent/Parent.js
// The MIT License (MIT)

// Copyright (c) 2015 Elliott Richerson

// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:

// The above copyright notice and this permission notice shall be included in 
// all copies or substantial portions of the Software.

// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

module.exports = function Parent(REST, db, validate, errors) {

    var schema = [{
        attribute: "name",
        type: "String",
        required: true,
        auto: false,
        test: validate.parent_name()
    }, {
        attribute: "email",
        type: "String",
        required: true,
        auto: false,
        test: validate.parent_email()
    }, {
        attribute: "facebookID",
        type: "String",
        required: true,
        auto: false,
        test: validate.parent_facebook_id()
    }, {
        attribute: "created_on",
        type: "Date",
        required: false,
        auto: true
    }, {
        attribute: "updated_on",
        type: "Date",
        required: false,
        auto: true
    }];

    var fxns = {
        post: db.parent_create.bind(db),
        getOne: db.parent_by_facebook_id.bind(db),
        getAll: db.parents_all.bind(db),
        put: db.parent_update.bind(db),
        del: db.parent_delete_by_id.bind(db)
    };
    var parentREST = new REST(fxns, schema, validate, errors, "Parent");

    var parent = {
        schema: schema,
        post: parentREST.post,
        get: parentREST.get,
        put: parentREST.put,
        del: parentREST.del
    }
    return parent;
};

File: bbsync/mobile/iOS/Baby Sync/Baby Sync/Parent.swift
//
//  Parent.swift
//  Baby Sync
//
//  Created by Elliott Richerson on 9/29/15.
//  Copyright © 2015 Elliott Richerson. All rights reserved.
//

import Foundation

struct Parent {
    var id: Int = 0
    var facebookID: Int = 0
    var name: String = ""
    var email: String = ""
    var createdOn: Date = Date()
    var updatedOn: Date = Date()
    
    internal var paramValue: Dictionary<String,AnyObject> {
        get {
            return ["facebookID":String(self.facebookID) as AnyObject,"name":self.name as AnyObject,"email":self.email as AnyObject]
        }
    }
    
    init() {
        self.id = 0
        self.facebookID = 0
        self.name = ""
        self.email = ""
        self.createdOn = Date()
        self.updatedOn = Date()
    }
    
    init(parent: JSON) {
        self.id = parent["id"].intValue
        self.facebookID = parent["facebookID"].intValue
        self.name = parent["name"].stringValue
        self.email = parent["email"].stringValue
        self.createdOn = ISO8601DateFormatter.sharedInstance.date(from: parent["created_on"].stringValue)!
        self.updatedOn = ISO8601DateFormatter.sharedInstance.date(from: parent["updated_on"].stringValue)!
    }
    
    init(parent: Parent) {
        self.id = parent.id
        self.facebookID = parent.facebookID
        self.name = parent.name
        self.email = parent.email
        self.createdOn = parent.createdOn
        self.updatedOn = parent.updatedOn
    }
}
